### PR TITLE
review: feature: CtElement#replace(Collection<CtElement>)

### PR DESCRIPTION
**Pull request by pvojtechovsky**
Support for replacing of one element by multiple.
Needed to implement more powerful `SubstitutionVisitor` ;-)

### CODE REVIEW

**Comment by pvojtechovsky** [comment]
Please review this one. 
What is problem of CI? Should I change something?

**Comment by monperrus** [comment]
interesting feature, what about a vararg here instead of a new interface method: `void replace(CtElement... element);`

File: src/main/java/spoon/generating/replace/ReplacementVisitor.java [hunk 1]
```diff
@@ -32,18 +34,24 @@
  *
  * This class is generated automatically by the processor {@link spoon.generating.ReplacementVisitorGenerator}.
  */
-class ReplacementVisitor extends CtScanner {
+class ReplacementVisitor<E extends CtElement> extends CtScanner {
```
Comment (monperrus) [inline src/main/java/spoon/generating/replace/ReplacementVisitor.java]: A new formal type parameter added to element: 'E extends spoon.reflect.declaration.CtElement'.
Why do you need this?

Comment (pvojtechovsky) [inline src/main/java/spoon/generating/replace/ReplacementVisitor.java]: Good question. I can live without that. I will remove it.

**Comment by pvojtechovsky** [comment]
> vararg method: void replace(CtElement... element);

nice! I will do so.

**Comment by pvojtechovsky** [comment]
It is done from my point of view

File: src/main/java/spoon/support/reflect/declaration/CtElementImpl.java [hunk 2]
```diff
@@ -177,7 +177,7 @@ public int hashCode() {
 
 	@Override
 	public void delete() {
-		replace(null);
+		replace();
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtElementImpl.java]: hard to understand. replace new CtElement[0] with a comment to explain?

**Comment by monperrus** [comment]
Hi Pavel,
I've just realized that introducing varargs is breaking at the binary level (and tested with the script below). Since `replace` is commonly used, this would be quite bad for client code ... and we have to undo you last commit. I'm sorry for my bad suggestion.

```
# test to check whether the introduction of varargs breaks binary compatibility

# without vararg
cat <<EOF > Foo.java
class Foo {
public void m(String arg){}
}
EOF

javac Foo.java

rm Foo.java

ls Foo.class

# now a client class
cat <<EOF > Bar.java
class Bar {
public static void main(String[] args){
new Foo().m("vv");
System.out.println("end");
}
}
EOF

javac Bar.java

java -cp . Bar


# with vararg

cat <<EOF > Foo.java
class Foo {
public void m(String... arg){}
}
EOF
javac Foo.java

# binary compatibility?
if java -cp . Bar; 
then echo binary compatible
else echo binary incompatible
fi 

# source compatibility after recompilation?
if javac Foo.java Bar.java;
then echo source compatible
else echo source incompatible
fi 


```

**Comment by pvojtechovsky** [comment]
no problem ;-) I will fix it tomorrow

**Pull request merged**